dlt/common/storages/fsspec_filesystem.py

```diff
@@ -52,39 +52,66 @@ class FileItem(TypedDict, total=False):
 
 
 def fsspec_filesystem(
-    protocol: str, credentials: FileSystemCredentials = None
+    protocol: str,
+    credentials: FileSystemCredentials = None,
+    kwargs: Optional[DictStrAny] = None,
+    client_kwargs: Optional[DictStrAny] = None,
 ) -> Tuple[AbstractFileSystem, str]:
     """Instantiates an authenticated fsspec `FileSystem` for a given `protocol` and credentials.
 
-    Please supply credentials instance corresponding to the protocol. The `protocol` is just the code name of the filesystem ie:
+    Please supply credentials instance corresponding to the protocol.
+    The `protocol` is just the code name of the filesystem i.e.:
     * s3
     * az, abfs
     * gcs, gs
 
     also see filesystem_from_config
     """
-    return fsspec_from_config(FilesystemConfiguration(protocol, credentials))
+    return fsspec_from_config(
+        FilesystemConfiguration(protocol, credentials, kwargs=kwargs, client_kwargs=client_kwargs)
+    )
 
 
 def fsspec_from_config(config: FilesystemConfiguration) -> Tuple[AbstractFileSystem, str]:
     """Instantiates an authenticated fsspec `FileSystem` from `config` argument.
 
-    Authenticates following filesystems:
+    Authenticates the following filesystems:
     * s3
     * az, abfs
     * gcs, gs
 
+    Additional fsspec filesystem arguments and client arguments are gathered from the
+    FilesystemConfiguration object and passed to the `url_to_fs` factory.
+
     All other filesystems are not authenticated
 
     Returns: (fsspec filesystem, normalized url)
 
     """
     proto = config.protocol
-    fs_kwargs: DictStrAny = {}
+
+    # defaults
+    fs_kwargs: DictStrAny = {"use_listings_cache": False}
+
+    if config.kwargs is not None:
+        fs_kwargs.update(config.kwargs)
+
+    if config.client_kwargs is not None:
+        fs_kwargs["client_kwargs"] = config.client_kwargs
+
     if proto == "s3":
-        fs_kwargs.update(cast(AwsCredentials, config.credentials).to_s3fs_credentials())
+        # we are careful not to override the client_kwargs key when updating fs_kwargs
+        s3_credentials = cast(AwsCredentials, config.credentials).to_s3fs_credentials()
+        if "client_kwargs" in fs_kwargs and "client_kwargs" in s3_credentials:
```

Comment:
would it not be possible to do the client_kwargs update after the if-clause thus having this code only once? you can extract the apropriate credentials dictionary for gcs also and use that. not super important, but might save some headache in future fs implementations

Reply:
I believe I simplified `fsspec_from_config` by moving the credential logic to `GcpCredentials`. The function became to unwieldy. Hopefully it is more readable now